cmd/upload.go

```diff
@@ -0,0 +1,155 @@
+	token    string
+	workflow string
+}
+
+// newUploadCmd creates a command to stop a running workflow.
+func newUploadCmd() *cobra.Command {
+	o := &uploadOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "upload",
+		Short: "Upload files and directories to workspace.",
+		Long:  uploadDesc,
+		Args:  cobra.ArbitraryArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return o.run(cmd, args)
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&o.token, "access-token", "t", "", "Access token of the current user.")
+	f.StringVarP(
+		&o.workflow,
+		"workflow",
+		"w", "",
+		"Name or UUID of the workflow. Overrides value of REANA_WORKON environment variable.",
+	)
+
+	return cmd
+}
+
+// TODO: filter files based on .gitignore and .reanaignore
+
+func (o *uploadOptions) run(cmd *cobra.Command, args []string) error {
+	var inputFiles, inputDirs []string
+
+	if len(args) > 0 {
+		// upload files and directories specified in arguments.
+		files, dirs, err := o.filterDirectories(args)
+		if err != nil {
+			return err
+		}
+		inputFiles, inputDirs = files, dirs
+	} else {
+		// upload all input files and directories specified in the reana.yaml file.
+		spec, err := workflows.GetWorkflowSpecification(o.token, o.workflow)
+		if err != nil {
+			return err
+		}
+		inputFiles = spec.Specification.Inputs.Files
+		inputDirs = spec.Specification.Inputs.Directories
+	}
+
+	files, err := o.collectFiles(inputFiles, inputDirs)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		_, err := workflows.UploadFile(o.token, o.workflow, file)
+		if err != nil {
+			return err
+		}
+		displayer.DisplayMessage(
+			fmt.Sprintf("File %s was successfully uploaded.", file),
+			displayer.Success,
+			false,
+			cmd.OutOrStdout(),
+		)
+	}
+	return nil
+}
+
+func (o *uploadOptions) collectFiles(files []string, dirs []string) ([]string, error) {
+	log.Debugf("Collect all the files which needs to be uploaded")
+	log.Debugf("files: %s", strings.Join(files, ", "))
+	log.Debugf("dirs: %s", strings.Join(dirs, ", "))
+	for _, dir := range dirs {
+		err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			// Check if a path is a symlink
+			if info.Mode()&os.ModeSymlink != 0 {
+				return nil
+			}
+			if !info.IsDir() {
+				files = append(files, path)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	log.Debugf("Collected files:")
+	log.Debugf("files: %s", strings.Join(files, ", "))
+	return files, nil
+}
+
+func (o *uploadOptions) filterDirectories(args []string) ([]string, []string, error) {
+	log.Debugf("Filter files and directories from arguments")
+	log.Debugf("args: %s", strings.Join(args, ", "))
+	var files, dirs []string
+	for _, path := range args {
+		pathInfo, err := os.Stat(path)
+		if err != nil {
+			return nil, nil, err
+		}
+		if pathInfo.IsDir() {
+			dirs = append(dirs, path)
+		} else {
+			files = append(files, path)
+		}
```

Comment:
Symlinks cause some troubles also here, as they can be appended to `files`.

Given that `filepath.Walk` works also on files, I am wondering if we can avoid dividing paths into `files` and `directories`. This would make the code less complicated, as we wouldn't need the `filterDirectories` function. On the other hand, if a user specifies a file in `directories` (or a directory in `files`) in the specification we would need to check manually whether that is actually a directory or not to spot the error. What do you think?

Reply:
(I also don't know if you would prefer fixing this issues later in a new PR)

Reply:
Good idea! I've changed the logic to use `filepath.Walk` on both files and directories and it works fine with less code, thanks! In this case even if user specifies a file in `directories` it works as expected.

>  On the other hand, if a user specifies a file in directories (or a directory in files) in the specification we would need to check manually whether that is actually a directory or not to spot the error. What do you think?

You mean we should report to a user if a file is found in `directories` and viceversa?

Reply:
> You mean we should report to a user if a file is found in `directories` and viceversa?

Yes, that is what I was referring to. I think we can discuss this in a separate issue, given that it also affects the python client!

Reply:
Yes, listing directories under `files` should be reported back to user, and vice versa. Since we have made that distinction, we should keep it.  

(Alternatively, we can respect both, so if a user uses only `files` and puts directories there, we can "silently" accept them. This is something that could pave the way towards using only one directive in the future.) 

BTW, regarding listing files and directories "unnecessarily", I have another observation. If a uses specifies the same file twice, it'll be uploaded twice, for example:

```
$ git diff
diff --git a/reana.yaml b/reana.yaml
index 4c69293..75d70e8 100644
--- a/reana.yaml
+++ b/reana.yaml
@@ -1,6 +1,11 @@
 version: 0.6.0
 inputs:
   files:
+    - code/gendata.C
+    - code/gendata.C
+    - code/gendata.C
+    - code/gendata.C
+    - code/gendata.C
     - code/gendata.C
     - code/fitdata.C
   parameters:
```

leads to:

```
$ reana-client run -w test
...
==> Uploading files...
==> Detected .gitignore file. Some files might get ignored.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/gendata.C was successfully uploaded.
==> SUCCESS: File /code/fitdata.C was successfully uploaded.
==> Starting workflow...
```

It would be nice to first make a list of all user-desired files and directories and walk through them, collect all the uploadable objects (avoiding symlinks), and then make a set out of them, so that we don't do any unnecessary upload the same object many times.

Reply:
P.S. The above upload behaviour was observed with REANA 0.8 clusters. Using the same client with REANA master branch clusters , one gets:

```
==> Starting workflow...
==> ERROR: Cannot start workflow test.1:
Input path declared multiple times: code/gendata.C
``` 

However, instead of throwing an error, it may be better to accept such a configuration and simply "deduplicate" the desired files and directories to be uploaded inside the client?

Reply:
(Especially because `reana-client validate` doesn't complain about the same file being listed multiple times. So, if we accept such a configuration as valid, we should allow processing it... Fully agree that these concerns are something for a new issue.)